### PR TITLE
gui: keep coin-control branches expanded across a directory re-slot

### LINKED ISSUE #3228: gui: scaling follow-ups for the windowed coin-selection model (#3183)
Follow-ups from the self-review of #3224 (the windowed coin-selection model for #3183). None of these are regressions — the legacy `QTreeWidget` behaved worse on every one of them — and none block that PR. They are the places where the new design's scaling story is incomplete, collected in one issue so they can be triaged together rather than rediscovered later.

Line references are against #3224 head `19eb564b7`.

## 1. Record removal is O(N) in the whole store

`GRC::CoinViews::applyRemove` locates the record by linear scan in `m_amount_order` (`findValue`) and then `shiftDown` walks **every** index in every view, every group's member list and the amount order to mirror the table compaction. `WalletCoinStore::removeRecordAt` additionally walks both identity maps (`m_by_outpoint`, `m_by_hash`) for the same shift.

So each removed record costs O(N) in the size of the whole store, and removals arrive in batches: spending a consolidation of 500 inputs on a 500k-coin wallet is ~500 × O(500k) of index maintenance under `cs_store`, which the GUI thread's reads block behind.

The append-and-compact table is what forces this — absolute indices change under everything that stores them. Options worth weighing:

- Stable slot ids with a free list instead of compaction (removal becomes O(log N) per index, no global shift), at the cost of a level of indirection on every read and a compaction pass when the free list grows.
- Tombstones plus a periodic rebuild, which keeps reads direct but makes `total` bookkeeping subtler.
- Keeping compaction but making the per-record locate O(log N) (position maps rather than `std::find`), which removes the scans but not the shift.

## 2. The group directory is fully materialized while children are windowed

Tree mode windows the children of each group but pulls the entire directory: `CoinSelectionModel::reseedFromSource` calls `getGroups(view, 0, -1)` and holds every `CoinGroupInfo` client-side, re-fetched on every Reset (a sort, a mode switch, a resync).

That is fine on the axis the acceptance gate exercises — many coins, few groups — but it is the opposite axis that hurts: a wallet with tens of thousands of distinct receiving addresses materializes tens of thousands of parent rows on every reseed. The `-devsyntheticcoins=<n>[:<groups>]` harness can already generate that shape, so the first step is measuring it (say `-devsyntheticcoins=50000:20000`) before deciding whether the directory needs the same `WindowCache` treatment the children get.

## 3. An expanded branch collapses whenever its parent re-slots

When a group's parent sort key moves, `CoinViews::reslotDirectory` emits `GroupRemove` + `GroupInsert` (correct — an in-place `GroupChange` would desynchronize the consumer's positional coordinates), and the model drops the realized child cache on the removal. The user-visible effect under the default amount sort: a coin arriving in an expanded mid-directory group collapses that branch and loses the scroll position.

The delta protocol is right; what's missing is consumer-side continuity — re-expanding and restoring the scroll offset for a group that was expanded before its row moved, keyed on the stable group address the model already maintains (`m_addr_id` / `m_id_addr`). Worth confirming how visible this is in practice before investing: on a quiet wallet a re-slot is rare, on a busy one it may be constant.

## 4. `CoinViews` enforces its invariants with `assert`

The positional core guards its container arithmetic with `assert` — the group exists, the member was found, `findGroup`/`findValue` returned something other than `npos` — and those are compiled out in release builds, which is what users run. A violated invariant that would abort a debug build instead becomes an out-of-range `erase`/`insert` on a `std::vector`.

The asserts are good documentation of the invariants and the unit tests exercise them, so this is not an argument for removing them. It is an argument for the handful that directly guard an index used in container arithmetic being early-returns as well, so a release build degrades to a stale view rather than memory corruption. `groupTouchDeltas` (added in #3224) already takes that shape and reads no worse.

---

Related: #3183 (the parent issue), #3224 (the PR these came out of), #3153 (the interfaces:: tracking issue).

### PR DESCRIPTION
**Pull request by PrestackI**
Addresses item 3 of #3228.

When a group's aggregate moves its sort key, `CoinViews::reslotDirectory` publishes `GroupRemove` + `GroupInsert` (`coinviews.cpp:312-316`) — correct, since an in-place `GroupChange` would desynchronize the consumer's positional coordinates. But the removal drops the realized child cache (`coinselectionmodel.cpp:751-757`) and QTreeView's expansion with it. Under the default amount sort, a coin arriving in an expanded mid-directory group therefore collapses that branch and loses the scroll position, though the user never collapsed anything.

Nothing puts it back today: `coinselectionview.cpp` has no re-expand, and `CoinControlDialog::expandPartiallySelected` (`coincontroldialog.cpp:204-228`) is wired only to load-finished, sort-click and tree-toggle — never to the event drain — and restores only `Qt::PartiallyChecked` parents.

## Change

The delta protocol is unchanged; this is consumer-side continuity.

The model now tracks the addresses the **user** has expanded separately from the realized slots in `m_groups`. That distinction is the whole mechanism: a re-slot drops the slot without the user having collapsed anything, so the two are otherwise indistinguishable. Branches are reported by stable id (`m_addr_id` / `m_id_addr`), which survives the further re-slots a deferred restore can race with, where a row number would not.

**The emission has to wait for the end of `applyCoinEventBatch`**, not fire from the insert handler. That handler holds a `StructureLock` across `begin/endInsertRows` (`:740-745`); `canFetchMore` answers false while `m_structure_locked` is nonzero (`:234-238`); and `QTreeView::expand()` calls `fetchMore()` synchronously. An expand issued inside the bracket marks the node expanded with the realization silently dropped — and nothing retries it, the node already being in QTreeView's expanded set. The view then defers once more through a `singleShot` continuation, the same discipline the collapse path already uses at `coinselectionview.cpp:31-46`, and restores the scroll offset it captured before expanding.

The connection is made in `CoinSelectionView::setModel` so both consumers (the coin-control dialog and the consolidation wizard page) get it, the way `uniformRowHeights` is set in the constructor rather than per dialog.

## Test

Adds `src/qt/test/coinselectionmodeltests.cpp` to `test_gridcoin-qt`, which `ctest` already runs in every `ENABLE_GUI` build. Five cases over the synthetic coin source: a re-slot of an expanded group is reported, a collapsed one is not, a real removal is not, a user collapse forgets the expansion, and the reported id resolves to the moved row. Reverting the model change turns the two positive cases red while the three negative controls stay green.

This needs a small test-only constructor taking `interfaces::WalletCoinSource&` directly. `WalletModel` supplies the model exactly two things — the display unit and the drain pump — and a test that injects batches through the existing public `applyCoinEventBatch` slot needs neither; four dereferences become null-guarded.

Scope note: the covered half is the model's. The view half — QTreeView re-expanding and the scroll restore — needs a `QApplication` and is not covered by these tests, so it is worth a look in the running dialog during review.

https://claude.ai/code/session_01Q5mqyoeyLygRtQTmUqPTTg

### CODE REVIEW

**Review by jamescowens — COMMENTED** [review]
Traced every anchor in the description. All of them hold, including the non-obvious ones.

| Claim | Confirmed at |
|---|---|
| A moved group is published as `GroupRemove` + `GroupInsert`, in place only when the position is unchanged | `wallet/coinviews.cpp:305-317` |
| `expandPartiallySelected` is wired only to load-finished, sort-click and tree-toggle | `coincontroldialog.cpp:194`, `:522`, `:532` — three call sites, nothing on the drain |
| ...and restores only `Qt::PartiallyChecked` | `coincontroldialog.cpp:215` |
| `canFetchMore` answers false while `m_structure_locked` is nonzero | `coinselectionmodel.cpp:272` |
| The emission lands outside every bracket | both `StructureLock`s are scoped inside their `if` blocks; none spans the function |

Two things I went looking for and did not find, which is the interesting part:

**The test seam is clean.** `m_wallet_model` has ten uses on the head. Six are inside the primary constructor, which always has a non-null model; the four reachable ones (`:375`, `:620`, `:634`, `:958`) are all guarded. The "four dereferences" count is exact and nothing was missed.

**"Realization IS expansion" holds.** `canFetchMore` returns false for an invalid parent (`:273`), so `fetchMore` can never run for the root and `m_expanded` cannot acquire an empty address. The only way in is a QTreeView expand of a group row.

The `m_expanded` / `m_groups` split is the right call, and so is deferring the emission to the end of the batch and then deferring *again* in the view. Reporting by stable id rather than row is the detail that makes the deferred restore survive a racing re-slot.

## Finding 1: the description's scope note is stale

> Scope note: the covered half is the model's. The view half — QTreeView re-expanding and the scroll restore — needs a `QApplication` and is not covered by these tests, so it is worth a look in the running dialog during review.

Commits 2 and 3 cover exactly that. `coinselectionviewtests.cpp` drives a real `QTreeView` under a real `QApplication` across five cases, including the scroll offset (`reslotHoldsTheScrollOffset`) and the `setModel` disconnect. The note reads like it was written against commit 1 and never revised, and it understates the PR to anyone deciding how carefully to hand-check the dialog.

Worth keeping the manual look regardless, but the description should say what is actually covered.

Incidentally `reslotHoldsTheScrollOffset` guards its own vacuity — `QVERIFY2(before > 0, "scrolled to the top, where a jump to the top is invisible")`. That is the right instinct and not common.

## Finding 2: `qt_import_plugins` is a new mechanism in a tree that already has one

The house pattern for static QPA plugins here is source-level `Q_IMPORT_PLUGIN`: `bitcoin.cpp:89-105` for the GUI, and `test/test_main.cpp:11-23` for this very binary, already covering minimal / xcb / windows / cocoa under `QT_STATICPLUGIN`. This PR adds offscreen through CMake `qt_import_plugins` instead — the first use of it anywhere in the repo.

I am **not** claiming the source-level route would have worked. I could not confirm that Qt defines a `QT_QPA_PLATFORM_OFFSCREEN` macro, and if it does not, the CMake route is necessary precisely for that reason. But that is the useful information, and it is not written down anywhere. As it stands the next person to touch this file sees two mechanisms doing one job and has a decent chance of "consolidating" them in the wrong direction.

A sentence in the CMake comment saying why the block at `test_main.cpp:11-23` was not simply extended would settle it permanently.

## Finding 3: the CMake comment's rationale is overstated, and this PR obsoletes it

> Every CI runner exports `QT_QPA_PLATFORM=offscreen` because it has no display

Four occurrences across two workflows, not every runner. More to the point, commit 2 adds a `qputenv("QT_QPA_PLATFORM", "offscreen")` default in `test_main.cpp`, so the binary now selects offscreen *itself* whether or not the environment says so. That is the real reason the plugin has to be linked, and it is a stronger reason than the one the comment gives — it holds for a developer running `ctest` locally on a machine that does have a display.

## Finding 4: `m_expanded` is never pruned on a genuine removal

The header says it is "pruned on a real collapse and cleared on Reset". The `GroupRemove` handler erases from `m_groups` (`:811`) but deliberately not from `m_expanded` — which is correct and load-bearing, since the removal half of a re-slot has to leave it intact for the insert half to find.

The consequence is that expansion is sticky *by address* for the model's lifetime. If a group is genuinely removed (last coin on it spent) and that address later returns as a genuinely new group in a later batch, it auto-expands, having never been expanded in that incarnation. By construction nothing can distinguish that from a re-slot, so I do not think it wants fixing — arguably it is even the friendlier behaviour.

But the header's list of prune points reads as exhaustive when it is missing the case that matters most for reasoning about the set's lifetime. One clause — that a removal deliberately does not prune, and why — would make it accurate.

## Minor

Two continuations can queue if `groupsReslotted` fires twice before the first `singleShot` turn, and the second captures an offset taken before the first expand ran. Purely cosmetic, and the comment already scopes the intent honestly ("enough to stop the jump to the top"). Noting it only so it is not mistaken for exact restoration later.

## Non-gating

Consumer-side entirely. The delta protocol is untouched — `coinviews.cpp` is unchanged in this diff, and nothing here reaches validation, the wallet's spend path, or the IPC schema. Two nodes cannot disagree about anything as a result of this.

CI is green at 25 checks.

ACK. Findings 1 and 3 are description/comment accuracy, 2 and 4 are single clarifying sentences. None of them touch the mechanism, which I think is right.

File: src/qt/coinselectionview.cpp [hunk 1]
```diff
@@ -46,6 +47,46 @@ CoinSelectionView::CoinSelectionView(QWidget* parent)
...
+    // parent at its new slot has already moved rows under the viewport, and
+    // re-expanding a branch above it moves them again. This holds the
+    // scrollbar still rather than reconstructing what was under the cursor --
+    // enough to stop the jump to the top, which is the visible symptom.
+    const int offset = verticalScrollBar()->value();
+    QTimer::singleShot(0, this, [this, group_ids, offset]() {
+        auto* m = qobject_cast<CoinSelectionModel*>(model());
+        if (!m) return;
+        for (const int id : group_ids) {
+            const QModelIndex idx = m->groupIndexForId(id);
```
[the remark this answers is not in the document]
Comment (jamescowens) [reply src/qt/coinselectionview.cpp]: Addressed in 46bf1f853: restoreReslottedBranches now pins a QPointer to the emitting model plus its registryGeneration() before queuing, and the continuation aborts on model destruction, a setModel() swap, or a generation bump (the generation is bumped in reseedFromSource, deliberately not in rebuildDirRows, so ordinary re-slots don't defeat the restore). Regression test: resetBetweenEmissionAndContinuationDoesNotReExpand.

File: src/qt/coinselectionmodel.cpp [hunk 2]
```diff
@@ -291,6 +327,9 @@ void CoinSelectionModel::releaseGroup(const QModelIndex& group_index)
     const GRC::CoinGroupInfo* g = groupAt(group_index);
     if (!g) return;
     const std::string address = g->address;
+    // The user collapsed it: forget the expansion even if the slot was never
+    // realized (an empty group has no cache to release).
+    m_expanded.erase(address);
```
[the remark this answers is not in the document]
Comment (jamescowens) [reply src/qt/coinselectionmodel.cpp:352]: Addressed in 788aa09fb: noteCollapsed() records the intent synchronously from the view's collapsed() handler (no bracket, no persistent index), and only the structural un-realize stays deferred. Regression test: userCollapseSurvivesAReslotInTheSameTurn reproduces exactly this ordering. Maintainer commit 9d21b410b extends the same discipline to the expand side (intent recorded from the expanded() signal rather than inferred inside fetchMore) and adds a fire-time isExpanded() re-validation to the deferred release.

File: src/qt/test/coinselectionmodeltests.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+//! in-place GroupChange would desynchronize the consumer's positional
+//! coordinates), and the removal drops the realized child cache along with
+//! QTreeView's expansion -- though the user never collapsed anything. The
+//! model has to tell that apart from a real collapse and from a real removal,
+//! and report the branches to restore by stable id.
+//!
+//! Driven through the model's test seam (a bare interfaces::WalletCoinSource,
+//! no WalletModel) over the synthetic coin source, with batches injected
+//! straight into applyCoinEventBatch. The view half -- QTreeView re-expansion
+//! and the scroll restore -- needs a QApplication and stays manual.
```
[the remark this answers is not in the document]
Comment (jamescowens) [reply src/qt/test/coinselectionmodeltests.h]: Addressed in 5025368cc: the note now points at CoinSelectionViewTests, which drives a real QTreeView on the offscreen platform.

File: src/qt/coinselectionmodel.cpp [hunk 4]
```diff
@@ -168,6 +193,14 @@ QModelIndex CoinSelectionModel::groupIndexByAddress(const std::string& address)
     return createIndex(row, 0, quintptr(0));
 }
 
+QModelIndex CoinSelectionModel::groupIndexForId(int group_id) const
+{
+    if (group_id < 0 || static_cast<std::size_t>(group_id) >= m_id_addr.size()) {
+        return QModelIndex();
+    }
+    return groupIndexByAddress(m_id_addr[static_cast<std::size_t>(group_id)]);
+}
```
[the remark this answers is not in the document]
Comment (jamescowens) [reply src/qt/coinselectionmodel.cpp:202]: False positive: groupIndexByAddress is declared const (coinselectionmodel.h:249) and defined const (coinselectionmodel.cpp:189). The branch compiles clean locally (Qt6 GUI + tests) and in CI.

File: src/qt/coinselectionview.cpp [hunk 5]
```diff
@@ -38,6 +59,12 @@ CoinSelectionView::CoinSelectionView(QWidget* parent)
...
             if (!pidx.isValid()) return;
+            // Re-validate at fire time: a re-expand landing after the
+            // collapse but before this turn (double-click, keyboard
+            // toggling, expandAll) means the user wants the branch open --
+            // un-realizing it now would leave a node the view shows as
+            // expanded with zero rows underneath.
+            if (isExpanded(pidx)) return;
             if (auto* m = qobject_cast<CoinSelectionModel*>(model())) {
                 m->releaseGroup(pidx);
             }
```
[the remark this answers is not in the document]
Comment (jamescowens) [reply src/qt/coinselectionview.cpp]: Fixed in 0d840a241: the collapse continuation now pins a QPointer to the model at collapse time and drops the continuation when the model died or is no longer the view's model, mirroring the guard restoreReslottedBranches already carries. Unreachable through today's dialogs (one view, one setModel call each), but the asymmetry between the two continuations was not worth keeping. Regression test modelSwapBeforeTheDeferredReleaseIsIgnored; severing the guard fails it.

File: src/qt/test/coinselectionviewtests.cpp [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+//! range; reslotHoldsTheScrollOffset asserts that it does before relying on it.
+constexpr int kViewWidth = 400;
+constexpr int kViewHeight = 160;
+
+//! The restore is deferred through a singleShot(0) continuation, so every
+//! assertion about it has to let the event loop turn first.
+void settle()
+{
+    QTest::qWait(50);
+}
```
[the remark this answers is not in the document]
Comment (jamescowens) [reply src/qt/test/coinselectionviewtests.cpp:37]: Declining the QTRY conversion. Most of these assertions are NEGATIVE (the branch must NOT re-expand, the restore must NOT fire) -- QTRY_VERIFY cannot express 'settle, then confirm nothing happened'; a bounded fixed wait is the correct tool there. The waited-on work is a singleShot(0) continuation, drained on the first event-loop turn, so the 50ms settle is ~three orders beyond what it needs even on a loaded host -- these tests complete in about a second total, and mixing two wait disciplines across positive and negative assertions in the same file costs more in reader load than the QTRY forms save.

File: src/qt/coinselectionmodel.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jamescowens) [reply src/qt/coinselectionmodel.cpp:202]: Repeat of the round-1 comment, adjudicated there as a false positive: groupIndexByAddress is const in both the declaration (coinselectionmodel.h:249) and the definition (coinselectionmodel.cpp:189), and the branch compiles clean locally and across the CI matrix.

File: src/qt/test/coinselectionviewtests.cpp [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
```
[the remark this answers is not in the document]
Comment (jamescowens) [reply src/qt/test/coinselectionviewtests.cpp:37]: Repeat of the round-1 comment; declined there. The assertions are mostly negative (nothing must happen after the settle), which QTRY_VERIFY cannot express, and the waited-on singleShot(0) work drains on the first event-loop turn, so the fixed 50ms wait is generous rather than tight.

**Review by jamescowens — APPROVED** [review]
Full adversarial review (three independent passes), four Copilot comments adjudicated across two converged rounds, and local verification: Qt6 GUI+tests build, full unit suite, model/view suites including the new regression tests, Fatal-mode QAbstractItemModelTester clean across all runs.

Two maintainer commits are included with PrestackI's maintainer-edit permission, both detailed in the review comments: 9d21b410b records expansion intent from the view's expanded() signal rather than inferring it in fetchMore() (fixing a same-turn collapse/re-expand desync that silently defeated the re-slot restore, and restoring compatibility with the COINSELECTION_MODEL_TESTER dev hook), and 0d840a241 pins the model across the deferred release continuation, mirroring the guard the restore continuation already carried. Both mutation-verified.

Full 27-check CI matrix green on this head.
